### PR TITLE
TEP-0121: Implement `Retries` in TaskRun

### LINKED ISSUE #5248: TaskRun retries: Improve separation of concerns between PipelineRun and TaskRun reconciler
Today, retries in Pipeline Tasks are implemented via the following process:
- If a TaskRun fails and the Pipeline Task has retries specified, the PipelineRun controller will clear the status of the TaskRun, add its old status to `taskRun.status.retriesStatus`, and mark it as running. (happens [here](https://github.com/tektoncd/pipeline/blob/b8e244a66b40da3d15e025eed92642461845ccee/pkg/reconciler/pipelinerun/pipelinerun.go#L810-L813))
- The TaskRun reconciler will use the length of its RetriesStatus to determine what to name the pod (happens [here](https://github.com/tektoncd/pipeline/blob/860b6285fe2494e9cb520d3b44ac6ada047aa836/pkg/pod/pod.go#L311-L313)

This pattern is a bit awkward because it means both reconcilers are partially responsible for implementing retries, and only the reconciler for a given CRD should be the one updating that CRD's status.

One way around this would be to create a new TaskRun for each retry instead. This was discussed as a potential implementation strategy when retries were created (in issue [221](https://github.com/tektoncd/pipeline/issues/221), initial PR [510](https://github.com/tektoncd/pipeline/pull/510), and final PR [658](https://github.com/tektoncd/pipeline/pull/658)), but it's not clear why the final decision was made.

If we'd like to go this route, here's what I think we should do:
- update the PipelineRun reconciler to create a new TaskRun for each retry, without changing the old TaskRun's status
- The new TaskRun can still have `RetriesStatus` based on the status of the old TaskRun.
- Because the new TaskRun will need a different name, it will create a different pod name anyway, so the logic around naming the pod based on the length of the retries status can be deleted.
- Now, RetriesStatus is solely informational. It serves no purpose to the TaskRun reconciler, so it can be removed from TaskRun.Status.

### LINKED ISSUE #5755: TEP-0121: Implement `Retries` in TaskRun reconciler

### PR DESCRIPTION
**Pull request by XinruZhang**
# Changes

This PR implements `Retries` in `TaskRun`, and removes the logic that PipelineRun controller relies on `RetriesStatus` to determine the termination of a TaskRun or CustomRun/Run.

Key Changes:
- New `Retries` field in both `v1beta1.TaskRun` and `v1.TaskRun`
- Archive retry attempt history in `RetriesStatus` for a failed `TaskRun`, before sending kubernetes and cloud events before a reconcile loop ends.
- Unit Tests to test the `TaskRun` object changes, especially the changes on `status.conditions` and `status.retriesStatus` after being reconciled once (one reconcile loop).
- [Integration Test](https://github.com/tektoncd/pipeline/blob/main/test/retry_test.go) still works.

fix #5755 
fix #5248
/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
API change: new `Retries` field in `v1beta1.TaskRun` and `v1.TaskRun`.
```

### CODE REVIEW

**Comment by JeromeJu** [comment]
Thanks @XinruZhang and thanks for making the comprehensive changes for the respective conversions.

**Review by JeromeJu — APPROVED** [review]
The api changes LGTM

File: pkg/apis/pipeline/v1/taskrun_types.go [hunk 1]
```diff
@@ -54,6 +54,9 @@ type TaskRunSpec struct {
 	// Status message for cancellation.
 	// +optional
 	StatusMessage TaskRunSpecStatusMessage `json:"statusMessage,omitempty"`
+	// Retries represents how many times this task run should be retried in case of task failure.
```
[the remark this answers is not in the document]
Comment (JeromeJu) [reply pkg/apis/pipeline/v1/taskrun_types.go:57]: very nit comment

[the remark this answers is not in the document]
Comment (XinruZhang) [reply pkg/apis/pipeline/v1/taskrun_types.go:57]: Thanks :)

File: pkg/reconciler/pipelinerun/pipelinerun_test.go [hunk 2]
```diff
@@ -3344,19 +3345,35 @@ func TestReconcileWithTimeoutAndRetry(t *testing.T) {
 	for _, tc := range []struct {
 		name               string
 		retries            int
+		retriesStatus      string
 		conditionSucceeded corev1.ConditionStatus
 		wantEvents         []string
 	}{{
-		name:               "One try has to be done",
-		retries:            1,
+		name:    "One try has to be done",
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:3352]: When I've encountered these tests before I've really struggled to understand what they do, and I think now they might even be a bit more confusing, since if retries are a taskrun feature why test them on a PipelineRun? Could you update the test comment and test names to clarify? I'm also not sure why `conditionSucceeded` is changing in the second test case.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 3]
```diff
@@ -196,50 +196,7 @@ func (t ResolvedPipelineTask) isFailure() bool {
...
-			if retriesDone < t.PipelineTask.Retries {
-				return true
-			}
-		}
-		return false
-	case t.IsCustomTask():
-		if t.Run == nil {
-			return true
-		}
-		retriesDone = len(t.Run.Status.RetriesStatus)
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:223]: This impacts custom tasks as well. Can the changes to custom tasks be split out into a separate PR?

File: pkg/apis/pipeline/v1/taskrun_types.go [hunk 4]
```diff
@@ -54,6 +54,9 @@ type TaskRunSpec struct {
 	// Status message for cancellation.
 	// +optional
 	StatusMessage TaskRunSpecStatusMessage `json:"statusMessage,omitempty"`
+	// Retries represents how many times this taskRun should be retried in case of task failure.
+	// +optional
+	Retries int `json:"retries,omitempty"`
 	// Time after which the build times out. Defaults to 1 hour.
```
Comment (lbernick) [inline pkg/apis/pipeline/v1/taskrun_types.go:60]: Let's also update the timeout docs to clarify that it's timeout per attempt

**Review by lbernick — COMMENTED** [review]
Closes https://github.com/tektoncd/pipeline/issues/5248

@XinruZhang could you test the behavior of your PR against https://github.com/tektoncd/pipeline/issues/4071#issuecomment-1241157709?

**Comment by XinruZhang** [comment]
> Closes #5248
> 
> @XinruZhang could you test the behavior of your PR against [#4071 (comment)](https://github.com/tektoncd/pipeline/issues/4071#issuecomment-1241157709)?

Yes it works as expected:
- TaskRun `hi` retries once, and fails on TimeOut on the second attempt. 
- TaskRun `bye` succeeds.
- PipelineRun fails on TimedOut.

```
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  creationTimestamp: "2022-11-29T16:38:48Z"
  generation: 2
  labels:
    app.kubernetes.io/managed-by: tekton-pipelines
    tekton.dev/memberOf: tasks
    tekton.dev/pipeline: please-say-bye-again
    tekton.dev/pipelineRun: please-say-bye-again-gpgb4
    tekton.dev/pipelineTask: hi
  name: please-say-bye-again-gpgb4-hi
  namespace: default
  ownerReferences:
  - apiVersion: tekton.dev/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: PipelineRun
    name: please-say-bye-again-gpgb4
    uid: ad62ddd9-eb41-4123-9b32-87aeee37d57d
  resourceVersion: "118235147"
  uid: 698a3ed1-90da-43d3-8753-e3950be1119f
spec:
  resources: {}
  retries: 2
  serviceAccountName: default
  status: TaskRunCancelled
  statusMessage: TaskRun cancelled as the PipelineRun it belongs to has timed out.
  taskSpec:
    steps:
    - image: alpine:3.12
      name: hi
      resources: {}
      script: |
        echo 'hi'
        sleep 10
        exit 1
  timeout: 1h0m0s
status:
  completionTime: "2022-11-29T16:39:07Z"
  conditions:
  - lastTransitionTime: "2022-11-29T16:39:07Z"
    message: TaskRun "please-say-bye-again-gpgb4-hi" was cancelled. TaskRun cancelled
      as the PipelineRun it belongs to has timed out.
    reason: TaskRunCancelled
    status: "False"
    type: Succeeded
  podName: please-say-bye-again-gpgb4-hi-pod-retry1
  retriesStatus:
  - completionTime: "2022-11-29T16:39:03Z"
    conditions:
    - lastTransitionTime: "2022-11-29T16:39:03Z"
      message: |
        "step-hi" exited with code 1 (image: "docker.io/library/alpine@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69"); for logs run: kubectl -n default logs please-say-bye-again-gpgb4-hi-pod -c step-hi
      reason: Failed
      status: "False"
      type: Succeeded
    podName: please-say-bye-again-gpgb4-hi-pod
    startTime: "2022-11-29T16:38:48Z"
    steps:
    - container: step-hi
      imageID: docker.io/library/alpine@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69
      name: hi
      terminated:
        containerID: containerd://4a4744d380847cde62a9118bb5f2ec22e88d83bd69831406a50024d6925afc6c
        exitCode: 1
        finishedAt: "2022-11-29T16:39:02Z"
        reason: Error
        startedAt: "2022-11-29T16:38:52Z"
    taskSpec:
      steps:
      - image: alpine:3.12
        name: hi
        resources: {}
        script: |
          echo 'hi'
          sleep 10
          exit 1
  startTime: "2022-11-29T16:39:03Z"
  steps:
  - container: step-hi
    imageID: docker.io/library/alpine@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69
    name: hi
    terminated:
      exitCode: 1
      finishedAt: "2022-11-29T16:39:07Z"
      reason: TaskRunCancelled
      startedAt: "2022-11-29T16:39:06Z"
  taskSpec:
    steps:
    - image: alpine:3.12
      name: hi
      resources: {}
      script: |
        echo 'hi'
        sleep 10
        exit 1
```

```
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  creationTimestamp: "2022-11-29T16:39:07Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: tekton-pipelines
    tekton.dev/memberOf: finally
    tekton.dev/pipeline: please-say-bye-again
    tekton.dev/pipelineRun: please-say-bye-again-gpgb4
    tekton.dev/pipelineTask: bye
  name: please-say-bye-again-gpgb4-bye
  namespace: default
  ownerReferences:
  - apiVersion: tekton.dev/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: PipelineRun
    name: please-say-bye-again-gpgb4
    uid: ad62ddd9-eb41-4123-9b32-87aeee37d57d
  resourceVersion: "118235223"
  uid: 7c8564bf-6d8b-4491-a323-8620cb690bde
spec:
  resources: {}
  serviceAccountName: default
  taskSpec:
    steps:
    - image: alpine:3.12
      name: bye
      resources: {}
      script: |
        echo 'bye'
  timeout: 1h0m0s
status:
  completionTime: "2022-11-29T16:39:12Z"
  conditions:
  - lastTransitionTime: "2022-11-29T16:39:12Z"
    message: All Steps have completed executing
    reason: Succeeded
    status: "True"
    type: Succeeded
  podName: please-say-bye-again-gpgb4-bye-pod
  startTime: "2022-11-29T16:39:07Z"
  steps:
  - container: step-bye
    imageID: docker.io/library/alpine@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69
    name: bye
    terminated:
      containerID: containerd://f54ec9f73202fe0b7fea7dc58db92fec7ca4bccaf06bdccaaf0fe0cad3acaafe
      exitCode: 0
      finishedAt: "2022-11-29T16:39:11Z"
      reason: Completed
      startedAt: "2022-11-29T16:39:11Z"
  taskSpec:
    steps:
    - image: alpine:3.12
      name: bye
      resources: {}
      script: |
        echo 'bye'
```

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  creationTimestamp: "2022-11-29T16:38:48Z"
  generateName: please-say-bye-again-
  generation: 1
  labels:
    tekton.dev/pipeline: please-say-bye-again
  name: please-say-bye-again-gpgb4
  namespace: default
  resourceVersion: "118235224"
  uid: ad62ddd9-eb41-4123-9b32-87aeee37d57d
spec:
  pipelineRef:
    name: please-say-bye-again
  serviceAccountName: default
  timeouts:
    pipeline: 1m0s
    tasks: 19s
status:
  completionTime: "2022-11-29T16:39:12Z"
  conditions:
  - lastTransitionTime: "2022-11-29T16:39:12Z"
    message: 'Tasks Completed: 2 (Failed: 1, Cancelled 0), Skipped: 0'
    reason: Failed
    status: "False"
    type: Succeeded
  finallyStartTime: "2022-11-29T16:39:07Z"
  pipelineSpec:
    finally:
    - name: bye
      taskSpec:
        metadata: {}
        spec: null
        steps:
        - image: alpine:3.12
          name: bye
          resources: {}
          script: |
            echo 'bye'
    tasks:
    - name: hi
      retries: 2
      taskSpec:
        metadata: {}
        spec: null
        steps:
        - image: alpine:3.12
          name: hi
          resources: {}
          script: |
            echo 'hi'
            sleep 10
            exit 1
  startTime: "2022-11-29T16:38:48Z"
  taskRuns:
    please-say-bye-again-gpgb4-bye:
      pipelineTaskName: bye
      status:
        completionTime: "2022-11-29T16:39:12Z"
        conditions:
        - lastTransitionTime: "2022-11-29T16:39:12Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        podName: please-say-bye-again-gpgb4-bye-pod
        startTime: "2022-11-29T16:39:07Z"
        steps:
        - container: step-bye
          imageID: docker.io/library/alpine@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69
          name: bye
          terminated:
            containerID: containerd://f54ec9f73202fe0b7fea7dc58db92fec7ca4bccaf06bdccaaf0fe0cad3acaafe
            exitCode: 0
            finishedAt: "2022-11-29T16:39:11Z"
            reason: Completed
            startedAt: "2022-11-29T16:39:11Z"
        taskSpec:
          steps:
          - image: alpine:3.12
            name: bye
            resources: {}
            script: |
              echo 'bye'
    please-say-bye-again-gpgb4-hi:
      pipelineTaskName: hi
      status:
        completionTime: "2022-11-29T16:39:07Z"
        conditions:
        - lastTransitionTime: "2022-11-29T16:39:07Z"
          message: TaskRun "please-say-bye-again-gpgb4-hi" was cancelled. TaskRun
            cancelled as the PipelineRun it belongs to has timed out.
          reason: TaskRunCancelled
          status: "False"
          type: Succeeded
        podName: please-say-bye-again-gpgb4-hi-pod-retry1
        retriesStatus:
        - completionTime: "2022-11-29T16:39:03Z"
          conditions:
          - lastTransitionTime: "2022-11-29T16:39:03Z"
            message: |
              "step-hi" exited with code 1 (image: "docker.io/library/alpine@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69"); for logs run: kubectl -n default logs please-say-bye-again-gpgb4-hi-pod -c step-hi
            reason: Failed
            status: "False"
            type: Succeeded
          podName: please-say-bye-again-gpgb4-hi-pod
          startTime: "2022-11-29T16:38:48Z"
          steps:
          - container: step-hi
            imageID: docker.io/library/alpine@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69
            name: hi
            terminated:
              containerID: containerd://4a4744d380847cde62a9118bb5f2ec22e88d83bd69831406a50024d6925afc6c
              exitCode: 1
              finishedAt: "2022-11-29T16:39:02Z"
              reason: Error
              startedAt: "2022-11-29T16:38:52Z"
          taskSpec:
            steps:
            - image: alpine:3.12
              name: hi
              resources: {}
              script: |
                echo 'hi'
                sleep 10
                exit 1
        startTime: "2022-11-29T16:39:03Z"
        steps:
        - container: step-hi
          imageID: docker.io/library/alpine@sha256:c75ac27b49326926b803b9ed43bf088bc220d22556de1bc5f72d742c91398f69
          name: hi
          terminated:
            exitCode: 1
            finishedAt: "2022-11-29T16:39:07Z"
            reason: TaskRunCancelled
            startedAt: "2022-11-29T16:39:06Z"
        taskSpec:
          steps:
          - image: alpine:3.12
            name: hi
            resources: {}
            script: |
              echo 'hi'
              sleep 10
              exit 1
```

File: docs/taskruns.md [hunk 5]
```diff
@@ -698,11 +699,25 @@ object(s), if present. Any `Request` or `Limit` specified by the user (on `Task`
...
+You can use the `retries` field to set how many times you want to retry on a failed TaskRun.
+All TaskRun failures are retriable except for `Cancellation`.
+
+For a retriable `TaskRun`, when an error occurs:
+- The error status is archived in `.status.RetriesStatus`
+- The `Succeed` condition in `.status` is updated: 
+```
+Type: Succeed
+Status: Unknown
+Reason: CompleteWithRetries
```
[the remark this answers is not in the document]
Comment (jerop) [reply docs/taskruns.md:712]: I'm also confused about this, but I think you mean that the TaskRun is still running but a retry was completed, is that right? If so, maybe reason `RetryCompleted` is clearer?

```yaml
Type: Succeed
Status: Unknown
Reason: RetryCompleted
```

We can even make it explicit by using reason `RetryFailed`

```yaml
Type: Succeed
Status: Unknown
Reason: RetryFailed
```

File: pkg/apis/pipeline/v1beta1/taskrun_types.go [hunk 6]
```diff
@@ -56,7 +56,10 @@ type TaskRunSpec struct {
 	// Status message for cancellation.
 	// +optional
 	StatusMessage TaskRunSpecStatusMessage `json:"statusMessage,omitempty"`
-	// Time after which the build times out. Defaults to 1 hour.
+	// Retries represents how many times this task run should be retried in case of task failure.
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/taskrun_types.go:59]:
```suggestion
	// Retries represents how many times this TaskRun should be retried in case of Task failure.
```

File: pkg/apis/pipeline/v1beta1/taskrun_types.go [hunk 7]
```diff
@@ -471,6 +474,15 @@ func (tr *TaskRun) IsTaskRunResultDone() bool {
 	return !tr.Status.GetCondition(apis.ConditionType(TaskRunConditionResultsVerified.String())).IsUnknown()
 }
 
+// IsRetriable returns true if the TaskRun's Retries is not exhausted.
+func (tr *TaskRun) IsRetriable() bool {
+	retriesDone := len(tr.Status.RetriesStatus)
+	if retriesDone < tr.Spec.Retries {
+		return true
+	}
+	return false
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/taskrun_types.go:483]:
```suggestion
	return len(tr.Status.RetriesStatus) < tr.Spec.Retries
```

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 8]
```diff
@@ -195,13 +196,15 @@ func (t ResolvedPipelineTask) isFailure() bool {
...
+		return isDone && c.IsFalse()
 	}
-	return isDone && c.IsFalse() && !t.hasRemainingRetries()
 }
 
-// hasRemainingRetries returns true only when the number of retries already attempted
-// is less than the number of retries allowed.
-func (t ResolvedPipelineTask) hasRemainingRetries() bool {
+// isCustomTaskRetriable returns true only when the number of retries already attempted
+// is less than the number of retries allowed for Custom Task Beta -- v1alpha1.Run.
```
Comment (jerop) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:204]:
```suggestion
// isCustomTaskRetriable returns true only when the number of retries already attempted
// is less than the number of retries allowed for v1alpha1.Run.
```

File: pkg/reconciler/taskrun/taskrun.go [hunk 9]
```diff
@@ -294,8 +295,20 @@ func (c *Reconciler) finishReconcileUpdateEmitEvents(ctx context.Context, tr *v1
...
-	events.Emit(ctx, beforeCondition, afterCondition, tr)
+	if afterCondition.IsFalse() && !tr.IsCancelled() && tr.IsRetriable() {
+		tr.Status.CompletionTime = &metav1.Time{Time: c.Clock.Now()}
+		newStatus := tr.Status.DeepCopy()
+		newStatus.RetriesStatus = nil
+		tr.Status.RetriesStatus = append(tr.Status.RetriesStatus, *newStatus)
+		tr.Status.StartTime = nil
+		tr.Status.PodName = ""
+		taskRunCondSet := apis.NewBatchConditionSet()
+		taskRunCondSet.Manage(&tr.Status).MarkUnknown(apis.ConditionSucceeded, "CompletedWithRetries", "")
```
Comment (jerop) [inline pkg/reconciler/taskrun/taskrun.go:306]: let's make the reason a constant `TaskRunReason` so it's easier to reuse (and change later)

https://github.com/tektoncd/pipeline/blob/83af16733e00b7572d2156e73b8266c71766a43e/pkg/apis/pipeline/v1beta1/taskrun_types.go#L155-L184

**Review by jerop — COMMENTED** [review]
Thank you @XinruZhang!

File: docs/taskruns.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (XinruZhang) [reply docs/taskruns.md:712]: And because we are using it as an intermediate status to signal that this TaskRun is going to be retried. I'm leaning towards `ToBeRetried`.

cc @pritidesai

[the remark this answers is not in the document]
Comment (jerop) [reply docs/taskruns.md:712]: Either `Retrying` or `ToBeRetried` sounds good to me, thanks for the clarification!

[the remark this answers is not in the document]
Comment (lbernick) [reply docs/taskruns.md:712]: +1 for `Retrying`!

[the remark this answers is not in the document]
Comment (vdemeester) [reply docs/taskruns.md:712]: Ah thank you @XinruZhang @jerop, it is clearer for me now. I do agree with either `Retrying` or `ToBeRetried` 🙃.

File: docs/taskruns.md [hunk 10]
```diff
@@ -698,11 +699,25 @@ object(s), if present. Any `Request` or `Limit` specified by the user (on `Task`
...
+```
+- `status.StartTime` and `status.PodName` are unset to trigger another retry attempt.
+
 ### Configuring the failure timeout
 
-You can use the `timeout` field to set the `TaskRun's` desired timeout value. If you do not specify this
-value for the `TaskRun`, the global default timeout value applies. If you set the timeout to 0, the `TaskRun` will
-have no timeout and will run until it completes successfully or fails from an error.
+You can use the `timeout` field to set the `TaskRun's` desired timeout value for each retry attempt. If you do
+not specify this value for the `TaskRun`, the global default timeout value applies. If you set the timeout to 0,
```
Comment (bendory) [inline docs/taskruns.md:719]: Please clarify how the global default timeout applies: does the global timeout apply _to each retry_ or does the global timeout apply _to the sum of all retries_?

File: pkg/apis/pipeline/v1beta1/swagger.json [hunk 11]
```diff
@@ -2793,6 +2793,11 @@
         "resources": {
           "$ref": "#/definitions/v1beta1.TaskRunResources"
         },
+        "retries": {
+          "description": "Retries represents how many times this TaskRun should be retried in case of Task failure.",
```
[the remark this answers is not in the document]
Comment (XinruZhang) [reply pkg/apis/pipeline/v1beta1/swagger.json]: Done.

File: docs/taskruns.md (hunk 10 above)
Comment (XinruZhang) [reply docs/taskruns.md]: SG~ updated to make it clearer!

Just to clarify, the doc here illustrates that the global timeout **value** applies, so the behavior is the same. But it's always good to be explicit.

**Comment by XinruZhang** [comment]
/retest
flake: TestReconcile/default_embedded_status

File: docs/taskruns.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (pritidesai) [reply docs/taskruns.md:712]: I have no preferences on the name and not good at it anyways 🙃 

This `status` describes the `taskRun` either exited with non-zero or timed out with `N` numbers of retries left.

[the remark this answers is not in the document]
Comment (jerop) [reply docs/taskruns.md:712]: @XinruZhang I see that you've updated the code to use reason `Retrying`, please update the docs here as well

[the remark this answers is not in the document]
Comment (XinruZhang) [reply docs/taskruns.md]: Update: I'm changing the `Reason` from `Retrying` to `ToBeRetried`. See https://github.com/tektoncd/community/pull/882#discussion_r1037209150 for detailed reason.

**Comment by XinruZhang** [comment]
/hold

One thing we haven't fully investigated is `Transient Errors`. We wouldn't want to retry a `TaskRun` when transient errors detected. For example: waiting for the remote resource to be ready, waiting for volumes to be mounted. Thanks @pritidesai  for pointing this out!

**Comment by XinruZhang** [comment]
/unhold

TL;DR: We do NOT and should NOT mark the resource as following for transient errors
```yaml
Type: Succeeded
Reason: xx
Status: False
```

The transient errors in taskrun implementation: once `tr.IsDone()` is true, the controller stops reconciling:

https://github.com/tektoncd/pipeline/blob/8c59d00342e9479e68e9c52f6d18b62ac3015489/pkg/reconciler/taskrun/taskrun.go#L127

And if the TaskRun failed to retrieve the remote Task, the taskRun marks the resource as `Ongoing`:

https://github.com/tektoncd/pipeline/blob/8c59d00342e9479e68e9c52f6d18b62ac3015489/pkg/reconciler/taskrun/taskrun.go#L336

For reference, in `PipelineRun`, it marks the PipelineRun status as `Running` when there's remote resource to be resolved, it's worth mentioning that PipelineRun controller tries to resolve those resources on each reconcile loop:

https://github.com/tektoncd/pipeline/blob/8c59d00342e9479e68e9c52f6d18b62ac3015489/pkg/reconciler/pipelinerun/pipelinerun.go#L526-L535

That's said, we won't mark the resource as Done (`Succeeded=False`) for transient errors.

**Comment by XinruZhang** [comment]
/retest
because of TestReconcile

File: pkg/apis/pipeline/v1beta1/taskrun_types.go [hunk 12]
```diff
@@ -471,6 +476,11 @@ func (tr *TaskRun) IsTaskRunResultDone() bool {
 	return !tr.Status.GetCondition(apis.ConditionType(TaskRunConditionResultsVerified.String())).IsUnknown()
 }
 
+// IsRetriable returns true if the TaskRun's Retries is not exhausted.
+func (tr *TaskRun) IsRetriable() bool {
+	return len(tr.Status.RetriesStatus) < tr.Spec.Retries
+}
+
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/taskrun_types.go:483]: please add unit tests for this function

File: pkg/apis/pipeline/v1beta1/taskrun_conversion.go [hunk 13]
```diff
@@ -85,6 +85,7 @@ func (trs *TaskRunSpec) ConvertTo(ctx context.Context, sink *v1.TaskRunSpec) err
 	}
 	sink.Status = v1.TaskRunSpecStatus(trs.Status)
 	sink.StatusMessage = v1.TaskRunSpecStatusMessage(trs.StatusMessage)
+	sink.Retries = trs.Retries
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/taskrun_conversion.go:88]: please add `retries` field to the `ConvertTo` tests in https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/taskrun_conversion_test.go

File: pkg/apis/pipeline/v1beta1/taskrun_conversion.go [hunk 14]
```diff
@@ -164,6 +165,7 @@ func (trs *TaskRunSpec) ConvertFrom(ctx context.Context, source *v1.TaskRunSpec)
 	}
 	trs.Status = TaskRunSpecStatus(source.Status)
 	trs.StatusMessage = TaskRunSpecStatusMessage(source.StatusMessage)
+	trs.Retries = source.Retries
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/taskrun_conversion.go:168]: please add `retries` field to the `ConvertFrom` tests in https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/taskrun_conversion_test.go

**Comment by XinruZhang** [comment]
Thanks @jerop! We've discussed offline, will break this PR into smaller pieces, and targeting to get all sub PRs merged before v0.43.0. Thanks!

**Comment by jerop** [comment]
Thank you for breaking it down @XinruZhang!
 
/hold

**Comment by XinruZhang** [comment]
Closing this PR because it is separeted into two sub prs: https://github.com/tektoncd/pipeline/pull/5842 and https://github.com/tektoncd/pipeline/pull/5844

**Comment by XinruZhang** [comment]
Closing this PR because it is separeted into two sub prs: https://github.com/tektoncd/pipeline/pull/5842 and https://github.com/tektoncd/pipeline/pull/5844
